core/src/main/scala/zio/jdbc/JdbcDecoder.scala

```diff
@@ -601,184 +603,174 @@ trait JdbcDecoderLowPriorityImplicits {
+      DynamicValue.Primitive(timestamp.toInstant(), StandardType.InstantType)
 
-      case SqlTypes.TIMESTAMP_WITH_TIMEZONE =>
-        // TODO: Timezone
-        val timestamp = resultSet.getTimestamp(columnIndex)
+    case SqlTypes.TIMESTAMP_WITH_TIMEZONE =>
+      // TODO: Timezone
+      val timestamp = resultSet.getTimestamp(columnIndex)
 
-        Some(DynamicValue.Primitive(timestamp.toInstant(), StandardType.InstantType))
+      DynamicValue.Primitive(timestamp.toInstant(), StandardType.InstantType)
 
-      case SqlTypes.TIME_WITH_TIMEZONE =>
-        // TODO: Timezone
-        val time = resultSet.getTime(columnIndex)
+    case SqlTypes.TIME_WITH_TIMEZONE =>
+      // TODO: Timezone
+      val time = resultSet.getTime(columnIndex)
 
-        Some(DynamicValue.Primitive(time.toLocalTime(), StandardType.LocalTimeType))
+      DynamicValue.Primitive(time.toLocalTime(), StandardType.LocalTimeType)
 
-      case SqlTypes.TINYINT =>
-        val short = resultSet.getShort(columnIndex)
+    case SqlTypes.TINYINT =>
+      val short = resultSet.getShort(columnIndex)
 
-        Some(DynamicValue.Primitive(short, StandardType.ShortType))
+      DynamicValue.Primitive(short, StandardType.ShortType)
 
-      case SqlTypes.VARBINARY =>
-        val chunk = getBinary(resultSet.getBinaryStream(columnIndex))
+    case SqlTypes.VARBINARY =>
+      val chunk = getBinary(resultSet.getBinaryStream(columnIndex))
 
-        Some(DynamicValue.Primitive(chunk, StandardType.BinaryType))
+      DynamicValue.Primitive(chunk, StandardType.BinaryType)
 
-      case SqlTypes.VARCHAR =>
-        val string = resultSet.getString(columnIndex)
+    case SqlTypes.VARCHAR =>
+      val string = resultSet.getString(columnIndex)
 
-        Some(DynamicValue.Primitive(string, StandardType.StringType))
+      DynamicValue.Primitive(string, StandardType.StringType)
 
-      case _ =>
-        None
-    }
   }
 
-  private def mapComplexType(resultSet: ResultSet, columnIndex: Int, sqlType: Int): Option[DynamicValue] =
-    sqlType match {
-      case SqlTypes.ARRAY =>
-        val arrayRs   = resultSet.getArray(columnIndex).getResultSet
-        val metaData  = arrayRs.getMetaData
-        var chunkData = Chunk.empty[DynamicValue]
-        while (arrayRs.next())
-          mapPrimitiveType(arrayRs, 2, metaData.getColumnType(2)).foreach { el =>
-            chunkData = chunkData :+ el
-          }
-        Some(DynamicValue.Sequence(chunkData))
-
-      // other SQL types
-      case _ =>
-        None
-    }
+  private def mapComplexType(resultSet: ResultSet, columnIndex: Int): PartialFunction[Int, DynamicValue] = {
+    case SqlTypes.ARRAY =>
+      val arrayRs  = resultSet.getArray(columnIndex).getResultSet
+      val metaData = arrayRs.getMetaData
+      val data     = mutable.ArrayBuffer.empty[DynamicValue]
+      while (arrayRs.next())
+        data += mapPrimitiveType(arrayRs, 2)(metaData.getColumnType(2))
```

Comment:
FYI, this is unsafe. If `mapPrimitiveType` matching doesn't match with the `metaData.getColumnType(2)` value, this will throw. Is it what you want?

Reply:
no, I want sql exception with some info. added applyOrElse there. thanks